### ISSUE TITLE
CUBLAS related error

### ISSUE THREAD

**Issue by DevSinghSachan** [clab/dynet#1057]
My code in dynet is getting the following error:
epoch:0.000673/01       train_loss:9.9675       lr:1.7469281488047272e-07
epoch:0.001345/01       train_loss:9.9751       lr:3.4938562976094545e-07
epoch:0.002018/01       train_loss:9.9434       lr:5.240784162197087e-07
epoch:0.002690/01       train_loss:9.9434       lr:6.987712595218909e-07
epoch:0.003363/01       train_loss:9.9434       lr:8.734640459806542e-07
epoch:0.004035/01       train_loss:9.9434       lr:1.0481568324394175e-06
epoch:0.004708/01       train_loss:9.9383       lr:1.2228496188981808e-06
CUBLAS failure in cublasSgemm(dev.cublas_handle, CUBLAS_OP_N, CUBLAS_OP_N, y.d.rows(), y.d.cols(), l.d.cols(), dev.kSCALAR_ONE, l.batch_ptr(b), l.d.rows(), r.batch_ptr(b), r.d.rows(), acc_scalar, y.batch_ptr(b), y.d.rows())
13
terminate called after throwing an instance of 'dynet::cuda_exception'
  what():  cublasSgemm(dev.cublas_handle, CUBLAS_OP_N, CUBLAS_OP_N, y.d.rows(), y.d.cols(), l.d.cols(), dev.kSCALAR_ONE, l.batch_ptr(b), l.d.rows(), r.batch_ptr(b), r.d.rows(), acc_scalar, y.batch_ptr(b), y.d.rows())

This happens on updating dynet to the latest version.
On one of my older version of the installed dynet, the error is not coming.

Thanks,

**Comment by chunyang-wen** [comment]
would please provide the information from
```shell
nvidia-smi
```

**Comment by DevSinghSachan** [comment]
Hi Chunyang,

Here is the output of `nvidia-smi` of both the systems I tried running the code on:

1 .)
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.44                 Driver Version: 367.44                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TIT...  Off  | 0000:02:00.0     Off |                  N/A |
| 22%   29C    P8    17W / 250W |   1182MiB / 12206MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX TIT...  Off  | 0000:03:00.0     Off |                  N/A |
| 22%   26C    P8    16W / 250W |      2MiB / 12206MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX TIT...  Off  | 0000:81:00.0     Off |                  N/A |
| 22%   26C    P8    17W / 250W |      2MiB / 12206MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce GTX TIT...  Off  | 0000:82:00.0     Off |                  N/A |
| 22%   47C    P2   101W / 250W |    893MiB / 12206MiB |     35%      Default |
+-------------------------------+----------------------+----------------------+

2) 
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 381.09                 Driver Version: 381.09                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 108...  Off  | 0000:02:00.0     Off |                  N/A |
| 23%   35C    P0    57W / 250W |      0MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 108...  Off  | 0000:03:00.0     Off |                  N/A |
|  0%   31C    P0    54W / 250W |      0MiB / 11172MiB |      2%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX 108...  Off  | 0000:82:00.0     Off |                  N/A |
| 41%   69C    P2   152W / 250W |   1858MiB / 11172MiB |     85%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce GTX 108...  Off  | 0000:83:00.0     Off |                  N/A |
| 30%   53C    P2    73W / 250W |   4961MiB / 11172MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

**Comment by DevSinghSachan** [comment]
I think this bug is related to dynet autobatching, as when I remove the `--dynet-autobatch 1` command line flag, the code works fine.

**Comment by chunyang-wen** [comment]
Thanks. Do you have the full stack trace?

**Comment by DevSinghSachan** [comment]
Actually, the job gets killed on the error. Apart from the above log, I don't get any logs. 

I am getting the error in this codebase:
https://github.com/DevSinghSachan/aiayn

To reproduce it, run the following command.

python train.py -s train-big.ja -t train-big.en --dynet-gpu --dynet-autobatch 1 --dynet-mem 10000

Also, let me know in case you need any more info.

**Comment by miguelballesteros** [comment]
I have exactly the same issue with a seq2seq with attention model.

**Comment by neubig** [comment]
Sorry about this, this is top priority when I get time to debug.

**Comment by neubig** [comment]
OK, I think this depends on the CUDA compiler/environment. I'm still looking into details, but I thought I'd post here in the time being. I ran the following command from the [dynet-benchmark repository](https://github.com/neulab/dynet-benchmark/):

    ./treenn-bulk --dynet-seed 2 --dynet-autobatch 1 32 32 0 2 2 600

and varied the following factors:

* `--dynet-autobatch 0` vs. `--dynet-autobatch 1`
* CPU vs. GPU
* older environment (gcc 4.8.4 + CUDA 7.5) vs. newer environment (gcc 5.4.0 + CUDA 9.0)

*only* in the case of where I picked the latter of three options (autobatching, GPU, and newer environment) I was able to get strange crashes. The variety of crash is inconsistent, so I imagine this is a memory error of some sort.
